pkg/controller/scyllaclustermigration/migrate_scyllacluster.go

```diff
@@ -0,0 +1,72 @@
+// Copyright (c) 2024 ScyllaDB.
+
+package scyllaclustermigration
+
+import (
+	"context"
+	"fmt"
+	"maps"
+
+	scyllav1 "github.com/scylladb/scylla-operator/pkg/api/scylla/v1"
+	"github.com/scylladb/scylla-operator/pkg/controller/scylladbdatacenter"
+	"github.com/scylladb/scylla-operator/pkg/controllerhelpers"
+	"github.com/scylladb/scylla-operator/pkg/naming"
+	"github.com/scylladb/scylla-operator/pkg/resourceapply"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func (scmc *Controller) syncScyllaDBDatacenter(ctx context.Context, sc *scyllav1.ScyllaCluster) ([]metav1.Condition, error) {
+	var progressingConditions []metav1.Condition
+
+	sdc, upgradeContext, err := migrateV1ScyllaClusterToV1Alpha1ScyllaDBDatacenter(sc)
+	if err != nil {
+		return progressingConditions, fmt.Errorf("can't migrate scyllav1.ScyllaCluster to v1alpha1.ScyllaDBDatacenter: %w", err)
+	}
+
+	existingSDc, err := scmc.scyllaDBDatacenterLister.ScyllaDBDatacenters(sc.Namespace).Get(sc.Name)
+	if err != nil && !errors.IsNotFound(err) {
+		return progressingConditions, fmt.Errorf("can't get v1alpha1.ScyllaDBDatacenter %q: %w", naming.ManualRef(sc.Namespace, sc.Name), err)
+	}
+
+	// Non-nil upgrade context means Operator was updated during upgrade of ScyllaCluster.
+	// Before we create the migrated ScyllaDBDatacenter, we have to make sure the ongoing upgrade state is migrated to new ConfigMap which holds it.
+	// Once ScyllaDBDatacenter is created, it becomes an authoritative upgrade state.
```

Comment:
Why don't we just generate the desired ScyllaDBDatacenter and desired ConfigMap and apply it?

- if ScyllaCluster has an upgrade state, it should be projected into the CM (but paused)
- if CM apply passes, delete the ScyllaCluster upgrade state
- if ScyllaCluster doesn't have an upgrade state, unpause the CM state machine, retain the existing state there (similar to say svcClusterIP)

Reply:
or we can pause it on the ScyllaDBDatacenter which would allow creating it before the CM but it feels like the split between the 2 resources complicates things